### PR TITLE
Add fr.sgued.ten-forward

### PR DESCRIPTION
**Pull request by sgued**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

### CODE REVIEW

File: flatpak-cargo-generator.py
Comment (bbhtt) [inline flatpak-cargo-generator.py:1]: Why is this here? If you want it to be in the same repo, add it as a git submodule.

File: ten-forward.git.json
Comment (bbhtt) [inline ten-forward.git.json:1]: Add it to the main manifest

File: flatpak-cargo-generator.py
Comment (sgued) [reply flatpak-cargo-generator.py:1]: I removed it

File: ten-forward.git.json
Comment (sgued) [reply ten-forward.git.json:1]: Done

File: fr.sgued.ten-forward.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        "CARGO_HOME": "/run/build/ten-forward/cargo"
+      }
+    },
+    "build-commands": [
+      "cargo --offline fetch --manifest-path Cargo.toml --verbose",
+      "cargo --offline build --release --verbose",
+      "install -Dm755 ./target/release/ten-forward -t /app/bin/"
+    ],
+    "post-install": [
+      "install -Dm755 dist/fr.sgued.ten-forward.metainfo.xml /app/share/appdata/fr.sgued.ten-forward.metainfo.xml",
```
Comment (bbhtt) [inline fr.sgued.ten-forward.json:33]: Fix these:

```
ppstream-util validate build/files/share/appdata/fr.sgued.ten-forward.appdata.xml 
build/files/share/appdata/fr.sgued.ten-forward.appdata.xml: FAILED:
• tag-missing           : <content_rating> required [use https://odrs.gnome.org/oars]
• tag-missing           : <release> required
Validation of files failed
```

File: fr.sgued.ten-forward.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
+{
+  "app-id": "fr.sgued.ten-forward",
```
Comment (bbhtt) [inline fr.sgued.ten-forward.json:2]: This should be `fr.sgued.ten_forward` https://freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-id-generic

File: fr.sgued.ten-forward.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      }
+    },
+    "build-commands": [
+      "cargo --offline fetch --manifest-path Cargo.toml --verbose",
+      "cargo --offline build --release --verbose",
+      "install -Dm755 ./target/release/ten-forward -t /app/bin/"
+    ],
+    "post-install": [
+      "install -Dm755 dist/fr.sgued.ten-forward.metainfo.xml /app/share/appdata/fr.sgued.ten-forward.metainfo.xml",
+      "install -Dm755 dist/fr.sgued.ten-forward.desktop /app/share/applacations/fr.sgued.ten-forward.desktop",
```
Comment (bbhtt) [inline fr.sgued.ten-forward.json:34]:
```suggestion
      "install -Dm755 dist/fr.sgued.ten-forward.desktop /app/share/applications/fr.sgued.ten-forward.desktop",
```

File: fr.sgued.ten-forward.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      "install -Dm755 assets/icon.svg /app/share/icons/hicolor/scalable/apps/fr.sgued.ten-forward.svg"
```
Comment (bbhtt) [inline fr.sgued.ten-forward.json:35]:
```suggestion
      "install -Dm644 assets/icon.svg /app/share/icons/hicolor/scalable/apps/fr.sgued.ten-forward.svg"
```

File: fr.sgued.ten-forward.json (hunk 1 above)
Comment (bbhtt) [inline fr.sgued.ten-forward.json:33]:
```suggestion
      "install -Dm644 dist/fr.sgued.ten-forward.metainfo.xml /app/share/appdata/fr.sgued.ten-forward.metainfo.xml",
```

File: fr.sgued.ten-forward.json (hunk 3 above)
Comment (bbhtt) [reply fr.sgued.ten-forward.json:34]: Change it to 644 too

Comment (bbhtt) [inline fr.sgued.ten-forward.json:34]:
```
desktop-file-validate fr.sgued.ten-forward.desktop 
fr.sgued.ten-forward.desktop: error: first group is not "Desktop Entry"
fr.sgued.ten-forward.desktop: error: file contains group "Desktop entry", but groups extending the format should start with "X-"
```

This makes appstream-compose fail

File: fr.sgued.ten-forward.json (hunk 2 above)
Comment (sgued) [reply fr.sgued.ten-forward.json:2]: Should i change it everywhere?

File: fr.sgued.ten-forward.json (hunk 3 above)
Comment (bbhtt) [reply fr.sgued.ten-forward.json:34]: It shows up at the very end of flatpak-builder output.

Comment (sgued) [reply fr.sgued.ten-forward.json:34]: I saw it initially for some errors with the metainfo file but it didn't show up for the desktop entry because of the type in the path that you noticed.

Comment (sgued) [reply fr.sgued.ten-forward.json]: I think it's fixed now. But then running manually `desktop-file-validate` on the entry raised some issues with categories that where not in the flatpak-builder output. I fixed them too.

File: fr.sgued.ten-forward.json (hunk 2 above)
Comment (sgued) [reply fr.sgued.ten-forward.json]: Done

**Comment by bbhtt** [comment]
bot, build fr.sgued.ten_forward

File: fr.sgued.ten-forward.json (hunk 3 above)
Comment (sgued) [reply fr.sgued.ten-forward.json]: Done

File: fr.sgued.ten-forward.json (hunk 1 above)
Comment (sgued) [reply fr.sgued.ten-forward.json]: Done

Comment (sgued) [reply fr.sgued.ten-forward.json]: Done

**Comment by sgued** [comment]
Is there a way to preview what it will look like (to validate the screenshots etc...) on flathub?

**Comment by bbhtt** [comment]
Screenshots won't work on test builds but you can install it and preview the appdata in GNOME Software/KDE Discover.

**Comment by sgued** [comment]
Thanks! All looks fine.

File: fr.sgued.ten-forward.json (hunk 2 above)
Comment (hfiguiere) [reply fr.sgued.ten-forward.json]: there was nothing wrong. the last component (application name) doesn't have the restrictions.

File: fr.sgued.ten_forward.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  "app-id": "fr.sgued.ten_forward",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "45",
+  "sdk": "org.gnome.Sdk",
+  "sdk-extensions": [
+    "org.freedesktop.Sdk.Extension.rust-stable"
+  ],
+  "command": "ten-forward",
+  "finish-args": [
+      "--env=RUST_LOG=info",
```
Comment (hfiguiere) [inline fr.sgued.ten_forward.json:11]: why do you need to set this?

File: fr.sgued.ten-forward.json (hunk 2 above)
Comment (bbhtt) [reply fr.sgued.ten-forward.json]: The spec seems to say any underscores in the id 

> ...replace any hyphens in the ID

Comment (hfiguiere) [reply fr.sgued.ten-forward.json]: > Well it does cause issues: [Relm4/relm4-template#26](https://github.com/Relm4/relm4-template/issues/26)

That's because it's no longer the last component.

Comment (hfiguiere) [reply fr.sgued.ten-forward.json]: > The spec seems to say any underscores in the id
> 
> > ...replace any hyphens in the ID

Not the flatpak documentation. But I guess when it is a hodgepodge of stubborn spec where people disagree... I mean they are the same the say to add `.desktop`.


Anyway, it is valid and it works, but I just wanted to point out it wasn't necessary.

Comment (bbhtt) [reply fr.sgued.ten-forward.json]: Hmm yea, I didn't notice that they differed here

File: fr.sgued.ten_forward.json (hunk 5 above)
Comment (sgued) [reply fr.sgued.ten_forward.json]: Removed

File: fr.sgued.ten-forward.json (hunk 2 above)
Comment (bbhtt) [reply fr.sgued.ten-forward.json]: I opened an issue for clarification https://github.com/ximion/appstream/issues/547
